### PR TITLE
[async] Fuse tasks only if they are either from the same kernel or arg-less

### PR DESCRIPTION
**Pull request by k-ye**
Also enabled mpm88 test :)

Related issue = N/A

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: tests/python/test_mpm88.py [hunk 1]
```diff
@@ -105,3 +104,14 @@ def substep():
...
   gui.update()
   '''
+
+
+@ti.all_archs
+def test_mpm88():
+    run_mpm88_test()
+
+
+@ti.run_async_test
```
Comment (archibate) [inline tests/python/test_mpm88.py:114]: Anti-#1462?

File: python/taichi/lang/__init__.py [hunk 2]
```diff
@@ -558,6 +558,10 @@ def wrapped(*test_args, **test_kwargs):
     return decorator
 
 
+def run_async_test(test):
+    return archs_with([cpu], async_mode=True)(test)
```
Comment (archibate) [inline python/taichi/lang/__init__.py:562]: Please do not add fuel to these ad-hoc test deco like `ti.all_archs` any more, they are to be removed in #1462.

File: taichi/program/async_engine.cpp [hunk 3]
```diff
@@ -257,12 +257,20 @@ bool AsyncEngine::fuse() {
...
+    auto task_b = rec_b.stmt;
     bool is_same_struct_for = task_a->task_type == OffloadedStmt::struct_for &&
                               task_b->task_type == OffloadedStmt::struct_for &&
                               task_a->snode == task_b->snode &&
                               task_a->block_dim == task_b->block_dim;
+    // TODO: a few problems with the range-for test condition:
+    // 1. This could incorrectly fuse two range-for kernels that have different
+    // sizes, but then the loop ranges get padded to the same power-of-two (E.g.
+    // maybe a side effect when a struct-for is demoted to range-for).
+    // 1. It has also fused range-fors that have the same linear range, but are
```
Comment (yuanming-hu) [inline taichi/program/async_engine.cpp:272]:
```suggestion
    // 2. It has also fused range-fors that have the same linear range, but are
```

**Comment by k-ye** [comment]
Let's not merge this until we can confirm if there's a problem on Win for running mpm88..

**Comment by k-ye** [comment]
It does seem like fusion runs super slow on Appveyor. Take `test_fuse_dense_x2y2z` as an example, many tests have passed during the execution of this single test case:

```
tests/python/test_fuse_dense.py::test_fuse_dense_x2y2z   <-- begin
[gw0] PASSED tests/python/test_dynamic.py::test_append_ret_value 
tests/python/test_dynamic.py::test_dense_dynamic 
[gw0] PASSED tests/python/test_dynamic.py::test_dense_dynamic 
tests/python/test_dynamic.py::test_dense_dynamic_len 
[gw0] PASSED tests/python/test_dynamic.py::test_dense_dynamic_len 
tests/python/test_dynamic.py::test_dynamic_activate 
[gw0] PASSED tests/python/test_dynamic.py::test_dynamic_activate 
tests/python/test_element_wise.py::test_binary_f[True-True] 
[gw0] PASSED tests/python/test_element_wise.py::test_binary_f[True-True] 
tests/python/test_element_wise.py::test_binary_f[True-False] 
[gw0] PASSED tests/python/test_element_wise.py::test_binary_f[True-False] 
tests/python/test_element_wise.py::test_binary_f[False-True] 
[gw0] PASSED tests/python/test_element_wise.py::test_binary_f[False-True] 
tests/python/test_element_wise.py::test_binary_i[is_mat0] 
[gw0] PASSED tests/python/test_element_wise.py::test_binary_i[is_mat0] 
tests/python/test_element_wise.py::test_binary_i[is_mat1] 
[gw0] PASSED tests/python/test_element_wise.py::test_binary_i[is_mat1] 
tests/python/test_element_wise.py::test_binary_i[is_mat2] 
[gw0] PASSED tests/python/test_element_wise.py::test_binary_i[is_mat2] 
tests/python/test_element_wise.py::test_writeback_binary_f[True] 
[gw0] PASSED tests/python/test_element_wise.py::test_writeback_binary_f[True] 
tests/python/test_element_wise.py::test_writeback_binary_f[False] 
[gw0] PASSED tests/python/test_element_wise.py::test_writeback_binary_f[False] 
tests/python/test_element_wise.py::test_writeback_binary_i[rhs_is_mat0] 
[gw0] PASSED tests/python/test_element_wise.py::test_writeback_binary_i[rhs_is_mat0] 
tests/python/test_element_wise.py::test_writeback_binary_i[rhs_is_mat1] 
[gw0] PASSED tests/python/test_element_wise.py::test_writeback_binary_i[rhs_is_mat1] 
tests/python/test_element_wise.py::test_unary 
[gw0] PASSED tests/python/test_element_wise.py::test_unary 
tests/python/test_element_wise.py::test_ternary_i[is_mat0] 
[gw0] PASSED tests/python/test_element_wise.py::test_ternary_i[is_mat0] 
tests/python/test_element_wise.py::test_ternary_i[is_mat1] 
[gw0] PASSED tests/python/test_element_wise.py::test_ternary_i[is_mat1] 
tests/python/test_element_wise.py::test_ternary_i[is_mat2] 
[gw0] PASSED tests/python/test_element_wise.py::test_ternary_i[is_mat2] 
tests/python/test_element_wise.py::test_ternary_i[is_mat3] 
[gw0] PASSED tests/python/test_element_wise.py::test_ternary_i[is_mat3] 
tests/python/test_element_wise.py::test_ternary_i[is_mat4] 
[gw0] PASSED tests/python/test_element_wise.py::test_ternary_i[is_mat4] 
tests/python/test_empty.py::test_empty 
[gw0] PASSED tests/python/test_empty.py::test_empty 
tests/python/test_empty.py::test_empty_args 
[gw0] PASSED tests/python/test_empty.py::test_empty_args 
tests/python/test_field.py::test_scalar_field[shape0-dtype0] 
[gw0] PASSED tests/python/test_field.py::test_scalar_field[shape0-dtype0] 
tests/python/test_field.py::test_scalar_field[shape0-dtype1] 
[gw0] PASSED tests/python/test_field.py::test_scalar_field[shape0-dtype1] 
tests/python/test_field.py::test_scalar_field[8-dtype0] 
[gw0] PASSED tests/python/test_field.py::test_scalar_field[8-dtype0] 
tests/python/test_field.py::test_scalar_field[8-dtype1] 
[gw0] PASSED tests/python/test_field.py::test_scalar_field[8-dtype1] 
tests/python/test_field.py::test_scalar_field[shape2-dtype0] 
[gw0] PASSED tests/python/test_field.py::test_scalar_field[shape2-dtype0] 
tests/python/test_field.py::test_scalar_field[shape2-dtype1] 
[gw0] PASSED tests/python/test_field.py::test_scalar_field[shape2-dtype1] 
tests/python/test_field.py::test_scalar_field[shape3-dtype0] 
[gw0] PASSED tests/python/test_field.py::test_scalar_field[shape3-dtype0] 
tests/python/test_field.py::test_scalar_field[shape3-dtype1] 
[gw0] PASSED tests/python/test_field.py::test_scalar_field[shape3-dtype1] 
tests/python/test_field.py::test_vector_field[shape0-dtype0-2] 
[gw0] PASSED tests/python/test_field.py::test_vector_field[shape0-dtype0-2] 
tests/python/test_field.py::test_vector_field[shape0-dtype0-3] 
[gw0] PASSED tests/python/test_field.py::test_vector_field[shape0-dtype0-3] 
tests/python/test_field.py::test_vector_field[shape0-dtype1-2] 
[gw0] PASSED tests/python/test_field.py::test_vector_field[shape0-dtype1-2] 
tests/python/test_field.py::test_vector_field[shape0-dtype1-3] 
[gw0] PASSED tests/python/test_field.py::test_vector_field[shape0-dtype1-3] 
tests/python/test_field.py::test_vector_field[8-dtype0-2] 
[gw0] PASSED tests/python/test_field.py::test_vector_field[8-dtype0-2] 
tests/python/test_field.py::test_vector_field[8-dtype0-3] 
[gw0] PASSED tests/python/test_field.py::test_vector_field[8-dtype0-3] 
tests/python/test_field.py::test_vector_field[8-dtype1-2] 
[gw0] PASSED tests/python/test_field.py::test_vector_field[8-dtype1-2] 
tests/python/test_field.py::test_vector_field[8-dtype1-3] 
[gw0] PASSED tests/python/test_field.py::test_vector_field[8-dtype1-3] 
tests/python/test_field.py::test_vector_field[shape2-dtype0-2] 
[gw0] PASSED tests/python/test_field.py::test_vector_field[shape2-dtype0-2] 
tests/python/test_field.py::test_vector_field[shape2-dtype0-3] 
[gw0] PASSED tests/python/test_field.py::test_vector_field[shape2-dtype0-3] 
tests/python/test_field.py::test_vector_field[shape2-dtype1-2] 
[gw0] PASSED tests/python/test_field.py::test_vector_field[shape2-dtype1-2] 
tests/python/test_field.py::test_vector_field[shape2-dtype1-3] 
[gw0] PASSED tests/python/test_field.py::test_vector_field[shape2-dtype1-3] 
tests/python/test_field.py::test_vector_field[shape3-dtype0-2] 
[gw0] PASSED tests/python/test_field.py::test_vector_field[shape3-dtype0-2] 
tests/python/test_field.py::test_vector_field[shape3-dtype0-3] 
[gw0] PASSED tests/python/test_field.py::test_vector_field[shape3-dtype0-3] 
tests/python/test_field.py::test_vector_field[shape3-dtype1-2] 
[gw0] PASSED tests/python/test_field.py::test_vector_field[shape3-dtype1-2] 
tests/python/test_field.py::test_vector_field[shape3-dtype1-3] 
[gw0] PASSED tests/python/test_field.py::test_vector_field[shape3-dtype1-3] 
tests/python/test_field.py::test_matrix_field[shape0-dtype0-1-2] 
[gw0] PASSED tests/python/test_field.py::test_matrix_field[shape0-dtype0-1-2] 
tests/python/test_field.py::test_matrix_field[shape0-dtype0-3-1] 
[gw0] PASSED tests/python/test_field.py::test_matrix_field[shape0-dtype0-3-1] 
tests/python/test_field.py::test_matrix_field[shape0-dtype0-4-3] 
[gw0] PASSED tests/python/test_field.py::test_matrix_field[shape0-dtype0-4-3] 
tests/python/test_field.py::test_matrix_field[shape0-dtype0-2-2] 
[gw0] PASSED tests/python/test_field.py::test_matrix_field[shape0-dtype0-2-2] 
tests/python/test_field.py::test_matrix_field[shape0-dtype1-1-2] 
[gw0] PASSED tests/python/test_field.py::test_matrix_field[shape0-dtype1-1-2] 
tests/python/test_image_io.py::test_image_io_uint[dt1-91-81-3-png] 
[gw0] PASSED tests/python/test_image_io.py::test_image_io_uint[dt1-91-81-3-png] 
tests/python/test_image_io.py::test_image_io_uint[dt2-91-81-3-png] 
[gw0] PASSED tests/python/test_image_io.py::test_image_io_uint[dt2-91-81-3-png] 
tests/python/test_immediate_layout.py::test_1D 
[gw0] PASSED tests/python/test_immediate_layout.py::test_1D 
tests/python/test_indices.py::test_indices 
[gw0] PASSED tests/python/test_indices.py::test_indices 
tests/python/test_indices.py::test_float_as_index 
[gw0] PASSED tests/python/test_indices.py::test_float_as_index 
tests/python/test_internal_func.py::test_basic 
[gw0] PASSED tests/python/test_internal_func.py::test_basic 
tests/python/test_internal_func.py::test_host_polling 
[gw0] PASSED tests/python/test_internal_func.py::test_host_polling 
tests/python/test_internal_func.py::test_list_manager 
[gw0] PASSED tests/python/test_internal_func.py::test_list_manager 
tests/python/test_internal_func.py::test_node_manager 
[gw0] PASSED tests/python/test_internal_func.py::test_node_manager 
tests/python/test_internal_func.py::test_node_manager_gc 
[gw0] PASSED tests/python/test_internal_func.py::test_node_manager_gc 
tests/python/test_kernel_template_mapper.py::test_kernel_template_mapper 
[gw0] PASSED tests/python/test_kernel_template_mapper.py::test_kernel_template_mapper 
tests/python/test_kernel_template_mapper.py::test_kernel_template_mapper_numpy 
[gw0] PASSED tests/python/test_kernel_template_mapper.py::test_kernel_template_mapper_numpy 
tests/python/test_kernel_templates.py::test_kernel_template_basic 
[gw0] PASSED tests/python/test_kernel_templates.py::test_kernel_template_basic 
tests/python/test_kernel_templates.py::test_kernel_template_gradient 
[gw1] PASSED tests/python/test_fuse_dense.py::test_fuse_dense_x2y2z   <-- done
```

**Pull request merged**